### PR TITLE
Bug 469636 - Add Smart Import support

### PR DESCRIPTION
**Pull request by Ian-Stewart-Binks**
This pull request adds [Smart Import](https://wiki.eclipse.org/E4/UI/Smart_Import) functionality for Buildship. The pull request adds the `org.eclipse.buildship.ui.importer` with the necessary files required by Smart Import, adds tests to the `org.eclipse.buildship.ui.test` module, and modifies three existing Buildship files: `ProjectImportJob.java`, `WorkspaceOperations.java` and `DefaultWorkspaceOperations.java`. The latter two files are changed in a way identical to changes made in [this pull request](https://github.com/eclipse/buildship/pull/100/files), with the purpose of exposing the nature adding functionality.

This pull request comes with an issue: due to the Smart Import dependency (`org.eclipse.e4.ui.importer`), this feature requires Eclipse Mars as the target platform, whereas Buildship supports platforms as far back as e36.

Note: Only Gradle projects with the Gradle wrapper are supported.

### CODE REVIEW

**Comment by eclipsewebmaster** [comment]
Issue tracker reference:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=469636

**Comment by etiennestuder** [comment]
Hi Ian. Thanks for the PR. We will look at it soon, once Mars is out and its tail is absorbed. We will most likely get back to you first before integration of the PR to better understand some of the decisions.
Regards, Etienne

**Comment by Ian-Stewart-Binks** [comment]
Sounds good, Etienne. Thanks!

**Comment by mickaelistria** [comment]
cc @oehme

**Comment by oehme** [comment]
Hey @Ian-Stewart-Binks and @mickaelistria. Sorry for not getting back to this in such a long time. We are still interested in smart import support and we've also done some simplifications to our synchronization code that should make it much easier to implement. 

If one of you is still interested in contributing, we'll be happy to support you.

File: org.eclipse.buildship.ui.import/plugin.xml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+<?eclipse version="3.4"?>
+<plugin>
+   <extension
+         point="org.eclipse.e4.ui.importer.projectConfigurators">
+      <projectConfigurator
+            class="org.eclipse.buildship.ui.smartimport.internal.GradleProjectConfigurator"
+            label="%gradle.importer.label">
+         <activeWhen>
+            <hasFile
+                  path="build.gradle">
```
Comment (eskatos) [inline org.eclipse.buildship.ui.import/plugin.xml]: This wouldn't work if the build script is named differently or is written in Kotlin and is named `build.gradle.kts`.
Relying on `settings.gradle` could become an issue too once it gets possible to write settings scripts in Kotlin in a `settings.gradle.kts` file.

**Comment by oehme** [comment]
Closing this for now, the API has significantly changed anyway. If someone is interested in providing a new PR, I'll be happy to review. Otherwise we'll prioritize this along other work.

**Pull request closed without merging**